### ISSUE TITLE
ng-packagr generates weird constructor for subclass which breaks angular dependency injection

### ISSUE THREAD

**Issue by deadalusai** [dherges/ng-packagr#679]
## Type of Issue

```
[X] Bug Report
[ ] Feature Request
```

## Description

Creating a subclass with an implicit constructor and initialized property causes the esm5 codegen to emit a weird form of the "pass-through" constructor.

#### How To Reproduce

1. Create a new empty library and prep it for `ng-packagr`
2. Add the following code:

    ```typescript
    export class Dependency { }
    export class Base {
        constructor (public dependency: Dependency) { }
    }
    export class SubBad extends Base {
        // This class generates a constructor which uses `super.apply(this, __spread(arguments))`
        property = 'value'
    }
    export class SubOk extends Base {
        // This class generates a constructor which uses `super.apply(this, arguments)`
    }
    ```
3. Run `ng-packagr`

4. Observe that the ems5 bundle file contains following JS for `SubBad` and `SubOk`:

    ```javascript
    var SubBad = /** @class */ (function (_super) {
        tslib_1.__extends(SubBad, _super);
        function SubBad() {
            var _this = _super.apply(this, tslib_1.__spread(arguments)) || this;
            // This class generates a constructor which uses `super.apply(this, __spread(arguments))`
            _this.property = 'value';
            return _this;
        }
        return SubBad;
    }(Base));
    var SubOk = /** @class */ (function (_super) {
        tslib_1.__extends(SubOk, _super);
        function SubOk() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return SubOk;
    }(Base));
    ```

5. Note the difference in the way the `super.apply` call is generated.

#### Expected Behaviour

The `super.apply` call should be the same for both constructors. This is currently necessary because Angular uses a regular expression (!) to determine if a constructor function is a pass-through, and the version which uses `tslib_1.__spread` fails this check. This bug https://github.com/dherges/ng-packagr/issues/631 then takes effect and Angular fails to construct the type correctly.

The regex check in Angular: https://github.com/angular/angular/blob/2fee5cc0958e08d16d6e549a056f3289138d5739/packages/core/src/reflection/reflection_capabilities.ts#L71

Angular should probably improve its code for detecting pass-through constructors, however it seems to be a non-trivial problem (it is hard to tell the difference between an zero-param constructor and a generated pass-through constructor). 

We could extend the regex in angular, however after passing through webpack the constructor is rewritten to look more like this:

```javascript
function SubBad () {
    var _this = _super.apply(this, __WEBPACK_IMPORTED_MODULE_0_tslib__["__spread"](arguments)) || this;
    _this.bordered = false;
    return _this;
}
```

...so maintaining the regex long-term seems like a losing battle.

#### Version Information

```
$ node_modules/.bin/ng-packagr --version
ng-packagr:            2.2.0
@angular/compiler:     5.2.8
@angular/compiler-cli: 5.2.8
rollup:                0.55.5
tsickle:               0.27.2
typescript:            2.7.2
```

**Comment by kurraraghu** [comment]
Hi ,
    I am facing similar issues after ng-packagr. Below i have defined my componet structure sample code. I am receiving **chartElement,zone,cd** properies as undefined. Can you provide solution to resolve this issue.
 
`@Component({
    selector: 'base-chart',
    template: `<div></div>`
})
export class BaseChartComponent implements OnChanges, AfterViewInit, OnDestroy {
    constructor(
        protected chartElement: ElementRef,
        protected zone: NgZone,
        protected cd: ChangeDetectorRef) {}
    ngOnDestroy(): void {
    }
    ngOnChanges(changes: SimpleChanges): void {
        this.update();
    }
    update(): void {
        -- -- -- --base component logic-- -
    }
}


@Component({
    selector: 'bar-vertical',
    template: `
    <svg>
       .........................
    <svg>
   `,
    changeDetection: ChangeDetectionStrategy.OnPush
    encapsulation: ViewEncapsulation.None
})
export class BarVerticalComponent extends BaseChartComponent {
    update(): void {
        super.update();
        const hostElem = this.chartElement.nativeElement;
//this.chartElement  is undefined
    }
}`

**Comment by deadalusai** [comment]
@kurraraghu The workaround is to add an explicit constructor to `BarVerticalComponent`:
```
export class BarVerticalComponent extends BaseChartComponent {
    constructor(
        chartElement: ElementRef,
        zone: NgZone,
        cd: ChangeDetectorRef
    ) {
        super(chartElement, zone, cd);
    }
}
```

**Comment by Swiftwork** [comment]
This should be resolved by [tsickle PR 760](https://github.com/angular/tsickle/pull/760) by @deadalusai, great work! The fix should be available in [tsickle v0.27.3](https://github.com/angular/tsickle/releases/tag/v0.27.3). We just need to update the dependency.

**Comment by dherges** [comment]
@deadalusai @Swiftwork thanks for the updates and the PRs!
